### PR TITLE
Add `BinarySerializer`, `EnumUtil::` and generator script

### PR DESCRIPTION
**Pull request by Maxxen**
This adds a `BinarySerializer` implementation to the new serialization infrastructure. It roughly follows the same format as the old one (using FieldWriters), but I've had to change some stuff around in how e.g ExtraTypeInfo and Vector serialization works.

I've also added a new `DeserializedStatementVerifierV2` pass that uses the new serialization code path. I'll try to remove it before the next release.

I've also added a `scripts/generate_enum_util.py` script that generates to/from string conversion code. although I had to add back some conversion switch-cases into the main code where the enum identifiers don't map 1:1 with their equivalents. Maybe we can extend the generator script to accept overrides in the future, but for now I think its good to keep it stupid so we don't end up too far in the meta programming swamp.

### CODE REVIEW

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great - good idea to have the enums be generated by a script!

I do think it's worthwhile to make sure all enum serialization code is consistent and everything can be generated by the script, either by adding manual overrides to the script for "legacy" enums, or by adjusting the enum serialization to the actual enum value. Having multiple different serializations of the same enum seems error prone and I would prefer if we can keep everything in the same place.

File: src/core_functions/scalar/list/list_sort.cpp [hunk 1]
```diff
@@ -244,14 +244,40 @@ static unique_ptr<FunctionData> ListSortBind(ClientContext &context, ScalarFunct
...
-static T GetOrder(ClientContext &context, Expression &expr) {
+static OrderType GetOrderType(ClientContext &context, Expression &expr) {
 	if (!expr.IsFoldable()) {
 		throw InvalidInputException("Sorting order must be a constant");
 	}
 	Value order_value = ExpressionExecutor::EvaluateScalar(context, expr);
 	auto order_name = StringUtil::Upper(order_value.ToString());
-	return EnumSerializer::StringToEnum<T>(order_name.c_str());
+	auto str = order_name.c_str();
+	if (StringUtil::Equals(str, "ORDER_DEFAULT") || StringUtil::Equals(str, "DEFAULT")) {
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply src/core_functions/scalar/list/list_sort.cpp:254]: In this case it gets complicated though, I assume that you want to be able to match on both "ASC" and "ASCENDING"?  How do roundtripping work if an enum has multiple string representations, which one do we use? 
In general I don't want the EnumUtil to have any knowledge of the rest of the systems invariants or SQL semantics, it should just be mechanical string<->enum conversions. But it seems to me there are parts of the system where that is not the case.

[the remark this answers is not in the document]
Comment (Mytherin) [reply src/core_functions/scalar/list/list_sort.cpp:254]: All we need is support for adding aliases to the `string -> enum` conversion code - I don't really see a problem with roundtripping being "lossy" with respect to enum aliases (e.g. `asc -> ascending`). Am I missing somewhere where that matters?

[the remark this answers is not in the document]
Comment (Maxxen) [reply src/core_functions/scalar/list/list_sort.cpp:254]: Alright, the less C++ I have to write the better.

**Comment by Maxxen** [comment]
@Mytherin Alright, what do you think of this?

**Comment by Maxxen** [comment]
Alright, so I couldn't get this merged because we got OOT extensions (subtstrait) which uses `JoinTypeToString`. I've added the helper functions back for now. We should really have some sort of `DEPRECATED` flag/attribute/annotation that issues a warning so we can better keep track of changes between dependent projects.

**Comment by Mytherin** [comment]
Thanks! Looks great.

> Alright, so I couldn't get this merged because we got OOT extensions (subtstrait) which uses `JoinTypeToString`. I've added the helper functions back for now. We should really have some sort of `DEPRECATED` flag/attribute/annotation that issues a warning so we can better keep track of changes between dependent projects.

A deprecated flag would be useful indeed - I don't know if we need to deprecate `JoinTypeToString` and friends, though. There is an ergonomic advantage to having the `ToString` method sit next to the original definition (don't need to include additional files) - and as long as `JoinTypeToString` calls into the `EnumUtil` there is no code duplication either.

**Pull request merged**